### PR TITLE
Rewrite `io::BufRead` doc examples to better demonstrate behaviors.

### PR DESCRIPTION
**Pull request by frewsxcv**
Prior to this commit, most of the `BufRead` examples used `StdinLock` to
demonstrate how certain `BufRead` methods worked. Using `StdinLock` is
not ideal since:

* Relying on run-time data means we can't show concrete examples of how
  these methods work up-front. The user is required to run them in order
  to see how they behave.
* If the user tries to run an example in the playpen, it won't work
  because the playpen doesn't support user input to stdin.

### CODE REVIEW

File: src/libstd/io/mod.rs [hunk 1]
```diff
@@ -1290,28 +1290,39 @@ pub trait BufRead: Read {
...
-    /// # Ok(())
-    /// # }
+    /// ```
+    /// use std::io::{self, BufRead};
+    ///
+    /// let mut cursor = io::Cursor::new(b"lorem-ipsum");
+    /// let mut buf = vec![];
+    ///
+    /// // cursor is at 'l'
+    /// let num_bytes = cursor.read_until(b'-', &mut buf).unwrap();
```
Comment (GuillaumeGomez) [inline src/libstd/io/mod.rs]: I'm not a big fan of `.unwrap()` in examples. Could you replace with `.expect()` please? Or you could even create a function which returns a `Result` and use `?` operator.

File: src/libstd/io/mod.rs [hunk 2]
```diff
@@ -1337,28 +1348,33 @@ pub trait BufRead: Read {
...
-    ///
-    ///     buffer.clear();
-    /// }
+    /// use std::io::{self, BufRead};
+    ///
+    /// let mut cursor = io::Cursor::new(b"foo\nbar");
+    /// let mut buf = String::new();
+    ///
+    /// // cursor is at 'f'
+    /// let num_bytes = cursor.read_line(&mut buf).unwrap();
```
Comment (GuillaumeGomez) [inline src/libstd/io/mod.rs]: Same.

**Review by GuillaumeGomez — DISMISSED** [review]

File: src/libstd/io/mod.rs (hunk 1 above)
Comment (frewsxcv) [inline src/libstd/io/mod.rs]: Since we're using a `Cursor` though, this will never fail. Should I still use `expect`? If so, what should the text say?

Comment (GuillaumeGomez) [inline src/libstd/io/mod.rs]: In this case add a comment explaining that `Cursor` will never fail. Both solutions are fine to me.

**Comment by GuillaumeGomez** [comment]
Great, thanks!

@bors: r+ rollup

**Pull request merged**